File: src/http-client-mock/UrlMatchers.cs
using System;
using System.Text.RegularExpressions;
using JetBrains.Annotations;

namespace HttpClientMock
{
    public static class UrlMatchers
    {
        [PublicAPI]
        public static Func<string, bool> Regex(string regexPattern)
        {
            return url =>
                {
                    var match = System.Text.RegularExpressions.Regex.Match(url, regexPattern, RegexOptions.IgnoreCase);
                    return match.Success;
                };
        }

        [PublicAPI]
        public static Func<string, bool> Is(string urlPattern)
        {
            return
                pathAndQuery =>
                {
                    var lowerStr = urlPattern.ToLower();
                    return pathAndQuery.ToLower() == (
                                        Uri.IsWellFormedUriString(lowerStr, UriKind.Absolute)
                                            ? new Uri(lowerStr).PathAndQuery
                                            : lowerStr
                                        );
                };
        }

        [PublicAPI]
        public static Func<string, bool> Wildcard(string wildCardPattern)
        {
            return s => new Regex($"^{System.Text.RegularExpressions.Regex.Escape(wildCardPattern)}$"
                .Replace("\\*", ".*")
                .Replace("\\?","."), RegexOptions.IgnoreCase).Match(s).Success;
        }
    }
}

File: src/fs-http-client-mock/UrlMatchers.fs
module HttpClientMock.UrlMatchers

open System;
open System.Text.RegularExpressions;

let Regex(regexPattern: string) =
    let f = fun url ->
        Regex.Match(url, regexPattern, RegexOptions.IgnoreCase).Success
    Func<string, bool>(f)

let Is (url: string) =
    let f = fun (pathAndQuery: string) ->
            let lowerStr = url.ToLower()
            pathAndQuery.ToLower() = match Uri.IsWellFormedUriString(lowerStr, UriKind.Absolute) with
                                        | true -> Uri(lowerStr).PathAndQuery
                                        | false ->lowerStr
    Func<string, bool>(f)
        
let Wildcard( wildCardPattern: string) =
    let regex = new Regex($"^{System.Text.RegularExpressions.Regex.Escape(wildCardPattern)}$"
            .Replace("\\*", ".*")
            .Replace("\\?","."), RegexOptions.IgnoreCase)
    Func<string, bool>(fun s -> regex.Match(s).Success)

